---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# healr

<!-- badges: start -->
<!-- badges: end -->

  *healr* provides function for the study of homeologous exchanges in neo-polyploids. It is a module of [HEAL](https://github.com/kenji-yt/HEAL), a snakemake workflow automating the main steps related to homeologous exchange estimation. While we recommend using *healr* after having processed your data with HEAL, you can use *healr* independantly from HEAL. 
  Functions from *healr* enable copy number inference at a subgenome level using subgenome classified bam-files (for example from  [eagle-rc](https://github.com/tony-kuo/eagle?tab=readme-ov-file#eagle-rc)). Then, inferred copy numbers can be combined with [GENESPACE](https://github.com/jtlovell/GENESPACE) output (synteny inference) to obtain copy number ratio for homeologous genes. The copy number information and ratio can be plotted and summarized in a number of way using *healr* functions.

## Bugs, Suggestions, Help 

If you wish to report an error/bug, ask for help, or suggest changes and new features, please open an issue [here](https://github.com/kenji-yt/healr/issues). 

## Installation

You can install the development version of healr from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("kenji-yt/healr", build_vignettes = TRUE)
```

## Usage 

Details on how to use *healr* can be found in the vignette of the package. You can see the vignette by running:

``` r
browseVignettes("healr")
```